src/plugins/dashboard/public/application/hooks/use_dashboard_app_state.ts

```diff
@@ -0,0 +1,351 @@
+  syncDashboardContainerInput,
+  savedObjectToDashboardState,
+  syncDashboardIndexPatterns,
+  syncDashboardFilterState,
+  loadSavedDashboardState,
+  buildDashboardContainer,
+  loadDashboardUrlState,
+  diffDashboardState,
+  areTimeRangesEqual,
+} from '../lib';
+
+export interface UseDashboardStateProps {
+  history: History;
+  savedDashboardId?: string;
+  redirectTo: DashboardRedirect;
+  isEmbeddedExternally: boolean;
+  kbnUrlStateStorage: IKbnUrlStateStorage;
+}
+
+export const useDashboardAppState = ({
+  history,
+  redirectTo,
+  savedDashboardId,
+  kbnUrlStateStorage,
+  isEmbeddedExternally,
+}: UseDashboardStateProps) => {
+  const dispatchDashboardStateChange = useDashboardDispatch();
+  const dashboardState = useDashboardSelector((state) => state.dashboardStateReducer);
+
+  /**
+   *  Dashboard app state is the return value for this hook and contains interaction points that the rest of the app can use
+   * to read or manipulate dashboard state.
+   */
+  const [dashboardAppState, setDashboardAppState] = useState<DashboardAppState>(() => ({
+    $onDashboardStateChange: new BehaviorSubject({} as DashboardState),
+    $triggerDashboardRefresh: new Subject<{ force?: boolean }>(),
+  }));
+
+  /**
+   *  Last saved state is diffed against the current dashboard state any time either changes. This is used to set the
+   * unsaved changes portion of the dashboardAppState.
+   */
+  const [lastSavedState, setLastSavedState] = useState<DashboardState>();
+  const $onLastSavedStateChange = useMemo(() => new Subject<DashboardState>(), []);
+
+  /**
+   * Unpack services
+   */
+  const services = useKibana<DashboardAppServices>().services;
+  const {
+    data,
+    core,
+    chrome,
+    embeddable,
+    indexPatterns,
+    usageCollection,
+    savedDashboards,
+    initializerContext,
+    savedObjectsTagging,
+    dashboardCapabilities,
+    dashboardSessionStorage,
+  } = services;
+  const { docTitle } = chrome;
+  const { notifications } = core;
+  const { query, search } = data;
+  const { getStateTransfer } = embeddable;
+  const { version: kibanaVersion } = initializerContext.env.packageInfo;
+
+  /**
+   * This useEffect triggers when the dashboard ID changes, and is in charge of loading the saved dashboard,
+   * fetching the initial state, building the Dashboard Container embeddable, and setting up all state syncing.
+   */
+  useEffect(() => {
+    // fetch incoming embeddable from state transfer service.
+    const incomingEmbeddable = getStateTransfer().getIncomingEmbeddablePackage(
+      DashboardConstants.DASHBOARDS_ID,
+      true
+    );
+
+    let canceled = false;
+    let onDestroy: () => void;
+
+    /**
+     * The dashboard build context is a collection of all of the services and props required in subsequent steps to build the dashboard
+     * from the dashboardId. This build context doesn't contain any extrenuous services.
+     */
+    const dashboardBuildContext: DashboardBuildContext = {
```

Comment:
I've changed the dashboard build context to include all required services unpacked, rather than as part of a `services` object. This means no more "restating the chain" in the functions down the line.